server/src/main/java/org/elasticsearch/action/admin/indices/datastream/DeleteDataStreamAction.java

```diff
@@ -165,10 +173,17 @@ static ClusterState removeDataStream(ClusterState currentState, Request request)
                 }
                 throw new ResourceNotFoundException("data_streams matching [" + request.name + "] not found");
             }
-            Metadata.Builder metadata = Metadata.builder(currentState.metadata());
+            List<String> dataStreamsToRemove = new ArrayList<>();
             for (String dataStreamName : dataStreams) {
                 logger.info("removing data stream [{}]", dataStreamName);
-                metadata.removeDataStream(dataStreamName);
+                DataStream dataStream = currentState.metadata().dataStreams().get(dataStreamName);
+                assert dataStream != null;
+                currentState = deleteIndexService.deleteIndices(currentState, new HashSet<>(dataStream.getIndices()));
```

Comment:
As a follow up we need to take into account that indices can be removed from a data stream (except the last/most current index). So we should either check here that the backing index really exists or in the delete index api also modify a data stream if the index is part of a data stream